### PR TITLE
Add support for Windows server 2019 using virtio-win-0.1.171 or later

### PR DESCRIPTION
**Pull request by felfert**
Ok, a clean PR now:
This PR adds support for Windows Server 2019, using latest virtio-win-0.1.171

While testing, I also found out, that the fallback to win2k12r2 drivers in WinImageBuilder.ps1, line 697 is probably a bad idea, because no error or warning is printed, but a non-functional image is generated.
This should be replaced by fatal error, clearly stating that the appropriate drivers are missing.

### CODE REVIEW

**Comment by felfert** [comment]
@ader1990 Did you had a chance to evaluate this yet?

**Comment by ader1990** [comment]
> @ader1990 Did you had a chance to evaluate this yet?

Hello, didn t get the chance yet, will come back with an answer next week.

File: WinImageBuilder.psm1 [hunk 1]
```diff
@@ -38,15 +38,16 @@ $VirtIODrivers = @("balloon", "netkvm", "pvpanic", "qemupciserial", "qxl",
...
+    "2k8" = @(60, 0, 1);
+    "2k8r2" = @(61, 0, 1);
+    "w7" = @(61, 0, 0);
+    "2k12" = @(62, 0, 1);
+    "w8" = @(62, 0, 0);
+    "2k12r2" = @(63, 0, 1);
+    "w8.1" = @(63, 0, 0);
+    "2k16" = @(100, 14393, 1);
+    "2k19" = @(100, 17763, 1);
+    "w10" = @(100, 0, 0);
```
Comment (ader1990) [inline WinImageBuilder.psm1]: there is no need to rewrite all these arrays, just add 2k19 key with 3 values and modify 2k16 with an additional value.

please try to keep backward compatibility in method params or arrays.

File: WinImageBuilder.psm1 [hunk 2]
```diff
@@ -668,6 +669,8 @@ function Get-VirtIODrivers {
         [parameter(Mandatory=$true)]
         [int]$MajorMinorVersion,
         [parameter(Mandatory=$true)]
+        [int]$BuildNumber,
```
Comment (ader1990) [inline WinImageBuilder.psm1]: all good here, as if we put buildnumber at the end, it will be quite messed up

File: WinImageBuilder.psm1 [hunk 3]
```diff
@@ -682,7 +685,10 @@ function Get-VirtIODrivers {
     foreach ($driver in $VirtioDrivers) {
         foreach ($osVersion in $VirtIODriverMappings.Keys) {
             $map = $VirtIODriverMappings[$osVersion]
-            if (!(($map[0] -eq $MajorMinorVersion) -and ($map[1] -eq $isServer))) {
+            if (!(($map[0] -eq $MajorMinorVersion) -and ($map[2] -eq $isServer))) {
```
Comment (ader1990) [inline WinImageBuilder.psm1]: here you are required to do a greater comparison, as 2016 or 2018 have more than one build number, it s rather a build number interval.

Comment (ader1990) [reply WinImageBuilder.psm1]: https://docs.microsoft.com/en-us/windows-server/get-started/windows-server-release-info

Comment (felfert) [reply WinImageBuilder.psm1]: Hmm, I don't think so, build numbers of **compatible** windows versions don't change. If the build number changes, then the driver might crash anyway. So it is safer to just check for equality.
In your link, you can see that easily: Windows Server 2019 and Windows Server version 1809 (the 2nd and 3rd Entry) have exactly the same build number.
People in [this](https://techcommunity.microsoft.com/t5/Windows-Server-Insiders/Windows-Server-2019-version-info/td-p/234472) discussion came to the same conclusion.

**Comment by felfert** [comment]
Also:
What do you think about my remark regarding the fallback to win2k12r2 drivers?
I would like to create a PR which replaces the fallback by a fatal error. Otherwise user could be fooled
if the wron drivers are installed and the resulting image crashes during boot.

File: WinImageBuilder.psm1 (hunk 3 above)
Comment (ader1990) [reply WinImageBuilder.psm1]: this way, you might end up not installing the virtio drivers for some windows versions. We can look at also the Get-ItemProperty -Path "HKLM:\SOFTWARE\Microsoft\Windows NT\CurrentVersion" -Name ReleaseId. As this tool should support intermediary (build) windows versions (test builds, etc), it is mandatory to check for the interval.

**Comment by ader1990** [comment]
> Also:
> What do you think about my remark regarding the fallback to win2k12r2 drivers?
> I would like to create a PR which replaces the fallback by a fatal error. Otherwise user could be fooled
> if the wron drivers are installed and the resulting image crashes during boot.

The fallback code is meant to install drivers from older virtio drivers on newer Windows version. For example, at the moment I made that fallback, there was Windows 10 released but the virtio stable drivers did only support Windows 8.1, that s why the fallback.

**Comment by ader1990** [comment]
we can make a fallback path work for 2019 too.

win10 and win16, afaik, can work with 2k12r2 drivers.

we can make a fallback path 2k16 to 2k12r2, 2k19 to 2k16. Or to remove it completely and fail the build which would be the sane thing to do, but people might want to install a known stable version of virtio of 2k12r2 for 2k16. Let me know what you think.

**Comment by felfert** [comment]
I would prefer the build to fail if drivers did not match. We can introduce a new option (e.g: enableVirtioDriverFallback) which is off by default. Then, people who know what they are doing can still enable the fallback at their own risk.

File: WinImageBuilder.psm1 (hunk 3 above)
Comment (ader1990) [reply WinImageBuilder.psm1]: I have checked the 0.1.171 and the files for 2k19 have the same hash as the files for 2k16. By default, the imaging tool code should install the 2k16 drivers for 2k19, which are the same. For the moment, there is no need for this change and I will come back with a refactored code when the stable virtio drivers contain different drivers for 2k19.

**Comment by ader1990** [comment]
> I would prefer the build to fail if drivers did not match. We can introduce a new option (e.g: enableVirtioDriverFallback) which is off by default. Then, people who know what they are doing can still enable the fallback at their own risk.

Fair enough, but the code should still work and install older drivers for newer versions, like it does currently. Maybe we can reuse the flag "force" to make this fallback.

**Comment by felfert** [comment]
The current fallback does *not* work with 2019.

**Comment by felfert** [comment]
I think, we should separate the 2 issues:
* Support for 2019 (this PR)
* Fallback to the correct drivers (not hardcoded to 2012) maybe in another PR.

With my original remark, I just wanted to express my opinion about the current fallback. which is suboptimal (in 2016 case) and not working at all (in the 2019 case). So why not simply merge **this** PR and **then** I create another PR containing an updated fallback (optional and to 2016 instead of 2012)

**Comment by zpeschke** [comment]
Is there an ETA for this? Thanks!

**Comment by mikeycarlton** [comment]
What's the current status on getting Server 2019 support out for the cloudbase imaging tools? Server 2019 has been out for quite some time now.

**Comment by ader1990** [comment]
@mikeycarlton @zpeschke imaging tools supported Windows Server 2019 out of the box.

Also, the latest stable Fedora virtio drivers for Windows 2019 are the same as the Windows 2016 drivers.

I am closing this PR and will create another one as soon as the Fedora VirtIO drivers will be different for 2019 with the suggested changes (the build number comparison should be forward looking, so that it supports future Windows Server 2019 releases).

Thank you,
Adrian Vladu

**Comment by mikeycarlton** [comment]
Awesome thanks! Could you update the Readme to reflect that it supports Server 2019?

**Comment by mikeycarlton** [comment]
Nevermind you beat me to it 👍

**Pull request closed without merging**